asQ/parallel_arrays.py

```diff
@@ -0,0 +1,206 @@
+from pyop2.mpi import MPI
+from numpy import zeros as zero_array
+
+
+def in_range(i, length, allow_negative=True, throws=False):
+    '''
+    Is the index i within the range of length?
+    :arg i: index to check
+    :arg length: the number of elements in the range
+    :arg throws: if True, then an IndexError is raised if the index is out of range
+    '''
+    if allow_negative:
+        result = (-length <= i < length)
+    else:
+        result = (0 <= i < length)
+    if throws and result is False:
+        raise IndexError(f"Index {i} is outside the range {length}")
+    return result
+
+
+class DistributedDataLayout(object):
+    def __init__(self, partition, comm=MPI.COMM_WORLD):
+        '''
+        A representation of a 1D set of data distributed over several MPI ranks.
+
+        :arg partition: The number of data elements on each rank. Can be a list of integers, in which case len(partition) must be comm.size. Can be a single integer, in which case all ranks have the same number of elements.
+        :arg comm: MPI communicator the data is distributed over.
+        '''
+        if isinstance(partition, int):
+            partition = [partition for _ in range(comm.size)]
+        else:
+            if len(partition) != comm.size:
+                raise ValueError(f"Partition size {len(partition)} not equal to comm size {comm.size}")
+        self.partition = partition
+        self.comm = comm
+        self.rank = comm.rank
+        self.local_size = partition[self.rank]
+        self.global_size = sum(partition)
+        self.offset = sum(partition[:self.rank])
+
+    def shift_index(self, i, itype='l', rtype='l'):
+        '''
+        Shift index between local and global addressing, and account for -ve indices.
+
```

Comment:
document the negative stuff a bit more please!

Reply:
See b5996c2